### PR TITLE
[RISCV] Break the (czero_eqz x, (setne x, 0)) -> x combine into 2 combines.

### PR DESCRIPTION
**Pull request by topperc**
We can think of this as two separate combines

(czero_eqz x, (setne y, 0)) -> (czero_eqz x, y)
and
(czero_eqz x, x) -> x

Similary the (czero_nez x, (seteq x, 0)) -> x combine can be broken into

(czero_nez x, (seteq y, 0)) -> (czero_eqz x, y)
and
(czero_eqz x, x) -> x

isel already does the (czero_eqz x, (setne y, 0)) -> (czero_eqz x, y) and (czero_nez x, (seteq y, 0)) -> (czero_eqz x, y) combines, but doing them early could expose other opportunities.

### CODE REVIEW

**Comment by topperc** [comment]
CC @zengdage

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 1]
```diff
@@ -16165,28 +16165,36 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
...
-    // czero_ne X, (xor Y, 1) -> czero_eq X, Y if Y is 0 or 1.
-    if (RHS.getOpcode() == ISD::XOR && isOneConstant(RHS.getOperand(1))) {
-      SDValue Cond = RHS.getOperand(0);
+    SDValue Val = N->getOperand(0);
+    SDValue Cond = N->getOperand(1);
+
+    unsigned Opc = N->getOpcode();
+
+    // czero_eqz x, x -> x
+    if (Opc && Val == Cond)
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16174]:
```suggestion
    if (Opc == RISCVISD::CZERO_EQZ && Val == Cond)
```

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 2]
```diff
@@ -16165,28 +16165,36 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
...
       APInt Mask = APInt::getBitsSetFrom(Cond.getValueSizeInBits(), 1);
-      if (DAG.MaskedValueIsZero(Cond, Mask)) {
-        unsigned NewOpc = N->getOpcode() == RISCVISD::CZERO_EQZ
-                              ? RISCVISD::CZERO_NEZ
-                              : RISCVISD::CZERO_EQZ;
-        return DAG.getNode(NewOpc, SDLoc(N), N->getValueType(0), LHS, Cond);
-      }
+      if (DAG.MaskedValueIsZero(Cond, Mask))
+        return DAG.getNode(InvOpc, SDLoc(N), N->getValueType(0), Val, Cond);
+    }
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16187]: Cond should be reset to `N->getOperand(1)`. Please add a test for `czero_eqz X, (xor (setcc y, 0, ne), 1)`.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp [hunk 3]
```diff
@@ -16165,28 +16165,36 @@ SDValue RISCVTargetLowering::PerformDAGCombine(SDNode *N,
...
+    // czero_eqz x, (setcc y, 0, ne) -> czero_eqz x, y
+    // czero_nez x, (setcc y, 0, ne) -> czero_nez x, y
+    // czero_eqz x, (setcc y, 0, eq) -> czero_nez x, y
+    // czero_nez x, (setcc y, 0, eq) -> czero_eqz x, y
+    if (Cond.getOpcode() == ISD::SETCC && isNullConstant(Cond.getOperand(1))) {
```
Comment (dtcxzyw) [inline llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16192]: I think the following logic can be removed now.

https://github.com/llvm/llvm-project/blob/3785d7424680e0bcb914a485af61be51559ab1ba/llvm/lib/Target/RISCV/RISCVInstrInfoZicond.td#L42-L49

**Review by dtcxzyw — CHANGES_REQUESTED** [review]

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 1 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16174]: Oops I refactored after I last ran tests.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 3 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16192]: Those can't remove removed. They handle cases other than RHS being 0.

File: llvm/lib/Target/RISCV/RISCVISelLowering.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16187]: DAG.MaskedValueIsZero will return true for that case because setcc produces 0/1.

Comment (dtcxzyw) [reply llvm/lib/Target/RISCV/RISCVISelLowering.cpp:16187]: At least we can use a new variable instead of overwriting Cond.

**Comment by dtcxzyw** [comment]
Hmm, this patch seems like a NFC(I) change :)

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Comment by topperc** [comment]
> Hmm, this patch seems like a NFC(I) change :) If not, please add some tests.

I don't think it's is quite NFC. Removing the setcc in DAGCombine can change use counts and enable other combines that may check that the setcc has only 1 use. I don't have a specific example in mind.

**Pull request merged**